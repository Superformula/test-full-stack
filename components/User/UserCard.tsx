import { ReactElement } from 'react'
import Link from 'next/link'
import { useRouter } from 'next/router'

import Avatar from '../generic/Avatar'
import LinkFunctionalChildWrapper from '../generic/LinkFunctionalChildWrapper'
import EditIcon from '../icons/EditIcon'
import TrashIcon from '../icons/TrashIcon'
import User from '../../models/user'
import { formatDate } from '../../utils/date-helpers'
import { parsePageQueryParam } from '../../utils/helpers'

import styles from './UserCard.module.css'

interface Props {
  user: User;
  onUserDelete: (userId: string) => Promise<void>;
}

export default function UserCard({ user, onUserDelete }: Props): ReactElement {
  const router = useRouter()
  const pageQueryParam = parsePageQueryParam(router.query)

  return (
    <div className={styles.wrapper}>
      <div className={styles.actionButtonsWrapper}>
        <span className={styles.actionButton} title="Edit User">
          <Link
            as={`/users/${user.id}?page=${pageQueryParam}`}
            href={`/?userId=${user.id}&page=${pageQueryParam}`}
            passHref
            scroll={false}
          >
            <LinkFunctionalChildWrapper>
              <EditIcon />
            </LinkFunctionalChildWrapper>
          </Link>
        </span>
        <span
          className={styles.actionButton}
          title="Delete User"
          onClick={(): Promise<void> => onUserDelete(user.id)}
        >
          <TrashIcon />
        </span>
      </div>
      <Avatar
        src={`https://source.unsplash.com/126x126/?portrait,face,${user.id}`}
        className={styles.avatar}
        alt={user.name}
      />
      <div className={styles.body}>
        <div className={styles.header}>
          <h3 className={styles.heading}>{user.name}</h3>
          <span className={styles.meta}>
            created{' '}
            <span className={styles.date}>{formatDate(user.createdAt)}</span>
          </span>
        </div>
        <p className={styles.description}>{user.description}</p>
      </div>
    </div>
  )
}
