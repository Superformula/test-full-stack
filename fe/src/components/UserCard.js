import React from 'react'
import { Card } from "react-bootstrap";

// --------------
// --------------
// --------------
const UserCard = ({ user }) => {
    // --------------
    return (
        <Card>
            <div style={{
                background: `transparent url('https://source.unsplash.com/collection/3053437/${user.id}') no-repeat 50% 50%`,
                width: "168px",
                height: "168px",
                borderRadius: "100%",
                backgroundSize: "cover"
            }}></div>

            <Card.Body>
                <Card.Title>{user.name}</Card.Title>
                <Card.Text>
                    Lorem ipsum dolor sit amet, consecteturâ€¦
                </Card.Text>
            </Card.Body>
        </Card>
    )
}

export default UserCard
