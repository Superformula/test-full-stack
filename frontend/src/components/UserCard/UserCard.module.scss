@import '../../sass/variables';

.userImage {
  width: 168px;
  height: 168px;
  border-radius: 50%;
}

.card{
  background: #ffffff;
  transition: box-shadow .2s ease-in-out;
  padding: 44px;
  border-radius: 8px;
  min-width: 19vw;

  &:hover {
    box-shadow: $box-shadow
  }

  h2 {
    max-width: 220px;
  }

  p {
    max-width: 200px;
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    display: block;
  }
}

.name {
  margin-top: 28px;
}