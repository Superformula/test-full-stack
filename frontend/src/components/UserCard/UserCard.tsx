import React from 'react';
import styles from './UserCard.module.scss';

type UserCardProps = {
  imageSrc: string;
  name: string;
  description: string;
};

const UserCard: React.FC<UserCardProps> = ({ imageSrc, name, description }) => (
  <div className={styles.card}>
    <img src={imageSrc} className={styles.userImage} alt="User avatar" />
    <h2 className={styles.name}>{name.toUpperCase()}</h2>
    <p>{description}</p>
  </div>
);

export default React.memo(UserCard);
