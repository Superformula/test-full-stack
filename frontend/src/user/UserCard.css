.user-card-container {
    padding: 25px 20px;
    margin: 20px 10px 20px 0;
    min-width: 280px;
    min-height: 336px;
    background-color: #ffffff;
    border-radius: 8px;
    position: relative;
}

.user-card-container:hover {
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
    transition: 0.3s;
}

.user-card-image {
    border-radius: 90px;
    margin: 0 auto;
    display: block;
}

.user-card-header {
    margin-bottom: 0;
    display: flex;
    justify-content: space-between;
}

.user-card-header p, h2 {
    margin: 15px 0;
}

.user-card-header h2 {
    font-style: normal;
    font-weight: 600;
    font-size: 21px;
    line-height: 26px;
}

.user-card-created-date {
    font-size: 16px;
    line-height: 32px;
    color: #c92020;
}

.user-card-description {
    overflow: hidden;
    text-overflow: ellipsis;
    display: -webkit-box;
    -webkit-line-clamp: 1; /* number of lines to show */
    -webkit-box-orient: vertical;
    margin-top: 0;
}

.user-card-header-container {
    position: absolute;
    bottom: 25px;
    left: 20px;
    width: 90%;
}

.user-card-pencil-icon-container {
    display: flex;
    justify-content: flex-end;
}

.user-card-pencil-icon {
    cursor: pointer;
}
