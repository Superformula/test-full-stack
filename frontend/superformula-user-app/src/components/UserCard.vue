<script src="__tests__/user-card.spec.js"></script>
<template>
  <div class="user-card">
    <button class="pencil" @click="edit"></button>
    <div class="user-image">
      <img :src="user.imageUrl" />
    </div>
    <div class="user-info">
      <h2 class="user-name">{{ user.name }}</h2>
      <p>created <span>{{ formatDate(user.createdAt) }}</span></p>
    </div>

    <p class="user-description">{{ user.description }}</p>
  </div>
</template>

<script>
import { shortStringFormat } from "../utils/date";

export default {
  name: 'UserCard',
  props: {
    user: {
      name: String,
      description: String
    }
  },
  methods: {
    edit() {
      this.$emit("edit");
    },
    formatDate(dateInput) { return shortStringFormat(dateInput); }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
 .user-card {
   background-color: #FFFFFF;
   width: 350px;
   height: 320px;
   position: relative;
   border-radius: 10px;
   margin: auto auto 60px;
 }

 .user-card:hover {
   box-shadow: 0px 8px 24px -10px rgba(0,0,0,0.75);
 }

 .user-card span {
   display: block;
 }

 .user-card:hover button {
   display: block;
 }

 .user-card:hover .user-info p {
   display: block;
   margin-bottom: 0;
 }

 .user-card button {
   position: absolute;
   top: 10px;
   right: 10px;
   display: none;
 }

 .user-image {
   display: inline-block;
   width: 170px;
   height: 170px;
   overflow: hidden;
   border-radius: 50%;
   margin-top: 25px;
 }

 .user-image img {
   height: 170px;
   margin-left: -25%;
 }

 .user-card:hover .user-name {
   width: 185px;
 }

 .user-name {
   font-size: 21px;
   text-align: left;
   margin-left: 20px;
   margin-bottom: 0;
   white-space: nowrap;
   height: 21px;
   width: 305px;
   overflow: hidden;
   text-overflow: ellipsis;
 }

 .user-description {
   font-family: 'Source Sans Pro', sans-serif;
   font-size: 16px;
   font-weight: 300;
   text-align: left;
   margin-left: 20px;
   margin-top: 5px;
   white-space: nowrap;
   height: 21px;
   width: 300px;
   overflow: hidden;
   text-overflow: ellipsis;
 }

 .user-info {
   height: 45px;
 }

 .user-info h2 {
   float:left;
   text-transform: uppercase;
 }

 .user-info p {
   float: right;
   margin-right: 20px;
   display: none;
 }

 .user-info p span {
   color: #AF1E1E;
   display: inline;
 }

 .pencil {
   background-image: url('../assets/pencil.png');
   background-size: 20px 20px;
   width: 20px;
   height: 20px;
   padding: 0;
   border: none;
   outline: none;
 }

 .pencil:hover {
   cursor: pointer;
   background-image: url('../assets/pencil-active.png');
 }
</style>
