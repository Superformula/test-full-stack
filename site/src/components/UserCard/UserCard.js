import styled from "styled-components";
import React from "react";
import moment from "moment/min/moment.min.js";
import { SfH2, SfP } from "../../styles/HtmlElementStyle.js";
import { SfPenIcon } from "../../utils/SfIcons.js";

const UserCardPanel = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
  flex-basis: 400px;
  background-color: white;
  height: 336px;
  margin: 32px;
  border-radius: 8px;
  :hover {
    box-shadow: 0 6px 9px 2px #e4e4e4;
  }
  svg {
    cursor: pointer;
    position: absolute;
    right: 23px;
    top: 18px;
  }
`;

const UserCardPanelImageArea = styled.div`
  display: flex;
  align-items: center;
  justify-items: center;
  justify-content: center;
  width: 100%;
  height: 243px;
  img {
    border-radius: 50%;
    width: 168px;
    height: 168px;
  }
`;

const UserCardPanelInfoArea = styled.div`
  font-size: 21px;
  height: 70px;
  margin-left: 32px;
  margin-right: 32px;
  .nameLabel {
    display: inline-block;
  }
  .descriptionLabel {
    padding-top: 8px;
    font-weight: 300;
    font-size: 16px;
  }

  .createdAtLabel {
    float: right;
    font-size: 16px;
    padding-top: 5px;
    p {
      display: inline;
    }
    label {
      margin-left: 5px;
      color: #b42320;
    }
  }
`;

const UserCard = (props) => {
  return (
    <UserCardPanel>
      <SfPenIcon
        onClick={() => {
          if (props.onClick) {
            props.onClick(props.user);
          }
        }}
      />
      <UserCardPanelImageArea>
        <img alt={props.user.name} src={props.user.image} />
      </UserCardPanelImageArea>
      <UserCardPanelInfoArea>
        <div>
          <SfH2 className="nameLabel">{props.user.name}</SfH2>
          <div className="createdAtLabel">
            <SfP> Created</SfP>
            <label>{moment(props.user.createdAt).format("YYYY-MM-DD")}</label>
          </div>
        </div>
        <div className="descriptionLabel">{props.user.description}</div>
      </UserCardPanelInfoArea>
    </UserCardPanel>
  );
};

export default UserCard;
