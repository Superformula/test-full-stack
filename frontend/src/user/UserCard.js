import React from 'react'
import './UserCard.css'

export const UserCard = ({ name, description, createdAt }) => {
  return (
    <div className={'user-card-container'}>
      <img className={'user-card-image'} src={'https://source.unsplash.com/user/erondu/180x180'}  alt={`user-${name}`}/>
      <div className={'user-card-header'}>
        <h2>{name}</h2>
        {createdAt && (
          <p className={'user-card-created-date'}>{`${createdAt}`}</p>
        )}
      </div>
      <p className={'user-card-description'}>{description}</p>
    </div>
  )
};
