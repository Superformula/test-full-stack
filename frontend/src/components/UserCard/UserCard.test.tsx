import React from 'react';
import { render } from '@testing-library/react';
import UserCard from './UserCard';

test('renders avatar and matches snapshot', () => {
  const { container } = render(<UserCard imageSrc="http://photo.com/photo" name="Pedro Yan" description="All good" />);
  expect(container).toMatchSnapshot();
});
