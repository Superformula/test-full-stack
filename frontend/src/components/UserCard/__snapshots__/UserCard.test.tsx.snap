// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders avatar and matches snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="User avatar"
      class="userImage"
      src="http://photo.com/photo"
    />
    <h2
      class="name"
    >
      PEDRO YAN
    </h2>
    <p>
      All good
    </p>
  </div>
</div>
`;
